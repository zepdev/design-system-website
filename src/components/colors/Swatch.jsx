import React, { Component } from 'react'
import PropTypes from 'prop-types'
import Button from '../ui/Button'
import CopyIcon from '../icons/CopyIcon'

class Swatch extends Component {
  state = {
    isTextCopied: false,
    copyError: false,
  }

  handleCopy = color => {
    if (document.queryCommandSupported('copy')) {
      let dummy = document.createElement('textarea')
      document.body.appendChild(dummy)
      dummy.value = color
      dummy.select()
      document.execCommand('copy')
      document.body.removeChild(dummy)
      this.setState({ isTextCopied: true })
      setTimeout(() => {
        this.setState({ isTextCopied: false })
      }, 2000)
      console.log(document.queryCommandSupported('copy'))
    } else {
      this.setState({ copyError: true })
      setTimeout(() => {
        this.setState({ copyError: false })
      }, 2000)
      console.log(document.queryCommandSupported('copy'))
    }
  }

  render() {
    const { title, color } = this.props
    const { isTextCopied, copyError } = this.state
    return (
      <div className="zds-swatch--container">
        <Button
          className="zds-swatch"
          style={{ background: color }}
          onClick={() => this.handleCopy(color)}
        >
          <span className="zds-swatch--backdrop" />
          <span className="zds-swatch--imageButton">
            {!isTextCopied && <CopyIcon className="zds-swatch--icon" />}
            <span className="zds-swatch--imageTitle" data-testid="swatch-text">
              {isTextCopied ? 'Copied!' : copyError ? 'Error!' : color}
            </span>
          </span>
        </Button>
        <p className="zds-swatch--title">{title}</p>
      </div>
    )
  }
}

Swatch.propTypes = {
  color: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
}

export default Swatch
