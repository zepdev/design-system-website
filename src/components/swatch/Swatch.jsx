import React, { useState } from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'
import withStyles from 'react-jss'
import ButtonBase from '../button/ButtonBase'
import CopyIcon from '../icons/CopyIcon'

const styles = theme => ({
  root: {
    display: 'flex',
    flexDirection: 'column',
    flexBasis: '48%',
    marginBottom: `${ theme.spacing.component.xxl.rem }rem`,
  },
  button: {
    height: '8rem',
    width: '100%',
    marginBottom: theme.spacing.component.s.px,
    position: 'relative',
    cursor: 'pointer',
    border: `1px solid ${ theme.colors.gray.grayMid.hex }`,
    '&:hover, &:focus': {
      zIndex: 1,
      '& $backdrop': {
        opacity: 0.4,
      },
      '& $backdropContent': {
        opacity: 1,
      },
    },
  },
  [`@media (min-width: ${ theme.breakpoints.s })`]: {
    root: {
      flexBasis: '22%',
    },
    button: {
      height: '8rem',
    },
  },
  [`@media (min-width: ${ theme.breakpoints.m })`]: {
    root: {
      marginRight: `${ theme.spacing.component.m.rem }rem`,
    },
  },
  [`@media (min-width: ${ theme.breakpoints.xxl })`]: {
    root: {
      marginRight: `${ theme.spacing.component.m.rem }rem`,
      flexBasis: '24%',
    },
    button: {
      height: '16rem',
    },
  },
  backdrop: {
    position: 'absolute',
    left: 0,
    right: 0,
    top: 0,
    bottom: 0,
    backgroundColor: theme.colors.gray.grayDark.hex,
    opacity: 0,
    transition: 0.5,
  },
  backdropContent: {
    position: 'absolute',
    left: 0,
    right: 0,
    top: 0,
    bottom: 0,
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    opacity: 0,
    color: theme.colors.gray.white.hex,
  },
  icon: {
    color: theme.colors.gray.white.hex,
    display: 'block',
  },
  buttonText: {
    position: 'relative,',
    paddingLeft: 5,
    display: 'block',
    textTransform: 'uppercase',
  },
  text: {
    margin: 0,
    color: theme.colors.gray.grayLight.hex,
    textTransform: 'uppercase',
  },
  name: {
    marginTop: 0,
    fontWeight: 500,
    marginBottom: theme.spacing.component.s.px,
  },
})

function Swatch({ color, classes }) {
  const [isTextCopied, setTextCopied] = useState(false)
  const [copyError, setCopyError] = useState(false)

  const handleCopy = color => {
    if (document.queryCommandSupported('copy')) {
      let dummy = document.createElement('textarea')
      document.body.appendChild(dummy)
      dummy.value = color
      dummy.select()
      document.execCommand('copy')
      document.body.removeChild(dummy)
      setTextCopied(true)
      setTimeout(() => {
        setTextCopied(false)
      }, 2000)
    } else {
      setCopyError(true)
      setTimeout(() => {
        setCopyError(false)
      }, 2000)
    }
  }

  return (
    <div className={classes.root}>
      <ButtonBase
        className={classes.button}
        style={{ background: color.hex }}
        onClick={() => handleCopy(color.hex)}
        data-testid="swatchButton"
      >
        <span className={classes.backdrop} />
        <span className={classes.backdropContent}>
          {!isTextCopied && (
            <CopyIcon className={classes.icon} ariaLabel={`copy hex:${ color.hex }`} />
          )}
          <span
            className={classnames(classes.buttonText, 'zep-typo--normal-4')}
            data-testid="swatchText"
          >
            {isTextCopied ? 'Copied!' : copyError ? 'Error!' : color.hex}
          </span>
        </span>
      </ButtonBase>
      <p className={classnames(classes.name, 'zep-typo--normal-2')}>{color.name}</p>
      <p className={classnames(classes.text, 'zep-typo--normal-1')}>{`HEX: ${ color.hex }`}</p>
      {color.rgb && (
        <p className={classnames(classes.text, 'zep-typo--normal-1')}>{`RGB: ${ color.rgb }`}</p>
      )}
      {color.hsb && (
        <p className={classnames(classes.text, 'zep-typo--normal-1')}>{`HSB: ${ color.hsb }`}</p>
      )}
    </div>
  )
}

Swatch.propTypes = {
  classes: PropTypes.object.isRequired,
  color: PropTypes.object.isRequired,
}

export default withStyles(styles)(Swatch)
