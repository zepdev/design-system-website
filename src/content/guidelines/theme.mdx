---
label: Guidelines
title: Theme
---

# Theme

Themes are used to customize existing elements to fit the specific style of a strategic business unit or product. There is no base color for the Zeppelin-element-library, almost all elements require that a theme class is present. At the moment the Zeppelin Design System offers three main themes. Developers can customize all elements by adding the theme class to any html tag (`<main>`,`<body>`,`<div>`, ...) as long it wraps all elements you want to use.

```js
<main class="theme-zeppelin" id="app-root">
  ...
</main>
```

We currently offer three main themes;

```js
// Zeppelin Indigo Blue
<div class="theme-zeppelin" >...</div>

// CAT Yellow
<div class="theme-cat" >...</div>

// Rental Red
<div class="theme-rental" >...</div>
```
