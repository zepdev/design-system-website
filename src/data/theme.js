module.exports = {
  breakpoints: {
    xs: '36em' /* 576px */,
    s: '48em' /* 768px */,
    m: '62em' /* 992px */,
    l: '75em' /* 1200px */,
    xl: '105em' /* 1680px */,
    xxl: '120em' /* 1920px */,
  },
  typography: {
    typefaces: {
      roboto: "'Roboto', Arial, Helvetica, sans-serif",
      teko: "Teko', sans-serif",
    },
    basics: {
      fontSizeBase: {
        specification: 'font-size-base',
        px: '16px',
        rem: '1rem',
        weight: null,
      },
      lineHeightBase: {
        specification: 'line-height-base',
        px: '24px',
        rem: '1.5rem',
        weight: null,
      },
      fontWeightBase: {
        specification: 'font-weight-base',
        px: null,
        rem: null,
        weight: 400,
      },
    },
    headings: {
      normal: {
        'zep-typo--normal-1': {
          rem: '0.75rem',
          px: '12px',
          weight: 400,
          specifications: [
            'font-family: "Roboto"',
            'font-weight: 400',
            'line-height: 1.5',
            'text-transform: "none"',
          ],
        },
        'zep-typo--normal-2': {
          rem: '0.75rem',
          px: '12px',
          weight: 400,
          specifications: [
            'font-family: "Roboto"',
            'font-weight: 400',
            'line-height: 1.5',
            'text-transform: "none"',
          ],
        },
        'zep-typo--normal-3': {
          rem: '0.75rem',
          px: '12px',
          weight: 400,
          specifications: [
            'font-family: "Roboto"',
            'font-weight: 400',
            'line-height: 1.5',
            'text-transform: "none"',
          ],
        },
        'zep-typo--normal-4': {
          rem: '0.75rem',
          px: '12px',
          weight: 400,
          specifications: [
            'font-family: "Roboto"',
            'font-weight: 400',
            'line-height: 1.5',
            'text-transform: "none"',
          ],
        },
        'zep-typo--normal-5': {
          rem: '0.75rem',
          px: '12px',
          weight: 400,
          specifications: [
            'font-family: "Roboto"',
            'font-weight: 400',
            'line-height: 1.5',
            'text-transform: "none"',
          ],
        },
        'zep-typo--normal-6': {
          rem: '0.75rem',
          px: '12px',
          weight: 400,
          specifications: [
            'font-family: "Roboto"',
            'font-weight: 400',
            'line-height: 1.5',
            'text-transform: "none"',
          ],
        },
        'zep-typo--normal-7': {
          rem: '0.75rem',
          px: '12px',
          weight: 400,
          specifications: [
            'font-family: "Roboto"',
            'font-weight: 400',
            'line-height: 1.5',
            'text-transform: "none"',
          ],
        },
        'zep-typo--normal-8': {
          rem: '0.75rem',
          px: '12px',
          weight: 400,
          specifications: [
            'font-family: "Roboto"',
            'font-weight: 400',
            'line-height: 1.5',
            'text-transform: "none"',
          ],
        },
      },
      display: {
        'zep-typo--display-1': {
          rem: '0.75rem',
          px: '12px',
          weight: 400,
          specifications: [
            'font-family: "Roboto"',
            'font-weight: 400',
            'line-height: 1.5',
            'text-transform: "none"',
          ],
        },
        'zep-typo--display-2': {
          rem: '0.75rem',
          px: '12px',
          weight: 400,
          specifications: [
            'font-family: "Roboto"',
            'font-weight: 400',
            'line-height: 1.5',
            'text-transform: "none"',
          ],
        },
        'zep-typo--display-3': {
          rem: '0.75rem',
          px: '12px',
          weight: 400,
          specifications: [
            'font-family: "Roboto"',
            'font-weight: 400',
            'line-height: 1.5',
            'text-transform: "none"',
          ],
        },
        'zep-typo--display-4': {
          rem: '0.75rem',
          px: '12px',
          weight: 400,
          specifications: [
            'font-family: "Roboto"',
            'font-weight: 400',
            'line-height: 1.5',
            'text-transform: "none"',
          ],
        },
      },
      special: {
        'zep-typo--special-1': {
          rem: '0.75rem',
          px: '12px',
          weight: 400,
          specifications: [
            'font-family: "Roboto"',
            'font-weight: 400',
            'line-height: 1.5',
            'text-transform: "none"',
          ],
        },
        'zep-typo--special-2': {
          rem: '0.75rem',
          px: '12px',
          weight: 400,
          specifications: [
            'font-family: "Roboto"',
            'font-weight: 400',
            'line-height: 1.5',
            'text-transform: "none"',
          ],
        },
        'zep-typo--special-3': {
          rem: '0.75rem',
          px: '12px',
          weight: 400,
          specifications: [
            'font-family: "Roboto"',
            'font-weight: 400',
            'line-height: 1.5',
            'text-transform: "none"',
          ],
        },
        'zep-typo--special-4': {
          rem: '0.75rem',
          px: '12px',
          weight: 400,
          specifications: [
            'font-family: "Roboto"',
            'font-weight: 400',
            'line-height: 1.5',
            'text-transform: "none"',
          ],
        },
      },
    },
  },
  colors: {
    primary: {
      indigoBlue: {
        name: 'indigo-blue',
        hex: '#27166f',
        rgb: '39/22/111',
        hsb: null,
      },
      catYellow: {
        name: 'cat-yellow',
        hex: '#f0b823',
        rgb: '240/184/35',
        hsb: null,
      },
      rentalRed: {
        name: 'rental-red',
        hex: '#af0e0e',
        rgb: '175/14/14',
        hsb: null,
      },
    },
    gray: {
      grayDark: {
        name: 'gray-dark',
        hex: '#101111',
        rgb: '16/17/17',
        hsb: null,
      },
      grayMidDark: {
        name: 'gray-mid-dark',
        hex: '#373d3f',
        rgb: '55/61/63',
        hsb: null,
      },
      gray: { name: 'gray', hex: '#464a4c', rgb: '70/74/76', hsb: null },
      grayLight: {
        name: 'gray-light',
        hex: '#6f7c80',
        rgb: '111/124/128',
        hsb: null,
      },
      grayLighter: {
        name: 'gray-lighter',
        hex: '#eceeef',
        rgb: '236/238/239',
        hsb: null,
      },
      grayMid: {
        name: 'gray-mid',
        hex: '#f2f3f4',
        rgb: '242/243/244',
        hsb: null,
      },
      grayLightest: {
        name: 'gray-lightest',
        hex: '#f7f7f9',
        rgb: '247/247/249',
        hsb: null,
      },
      white: { name: 'white', hex: '#ffffff', rgb: '255/255/255', hsb: null },
    },
    support: {
      alert: { name: 'alert', hex: '#ff6562', rgb: '255/101/98', hsb: null },
      warning: {
        name: 'warning',
        hex: '#f0ad4e',
        rgb: '240,/73/78',
        hsb: null,
      },
      success: { name: 'success', hex: '#5cb85c', rgb: '92/184/92', hsb: null },
      info: { name: 'info', hex: '#5bc0de', rgb: '91/192/222', hsb: null },
    },
  },
  spacing: {
    xxs: 4,
    xs: 8,
    s: 12,
    m: 16,
    l: 24,
    xl: 32,
    xxl: 48,
  },
}
