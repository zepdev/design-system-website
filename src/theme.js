import { createMuiTheme } from '@material-ui/core/styles'
import guidelines from '@zlab-de/zel-react/guidelines.json'

const theme = createMuiTheme({
  palette: {
    type: 'light',
    primary: {
      light: guidelines.theme.indigo.secondary,
      main: guidelines.theme.indigo.primary,
      dark: guidelines.theme.indigo.primary,
    },
    error: { main: guidelines.color.support.alert },
  },
  status: {
    gray: '#eceeef',
  },
  logo: guidelines.logo,
  blue: guidelines.theme.blue,
  indigo: guidelines.theme.indigo,
  red: guidelines.theme.red,
  yellow: guidelines.theme.yellow,
  guidelines: guidelines,
  space: guidelines.spacing.component,
  color: guidelines.color,
})

export default theme
