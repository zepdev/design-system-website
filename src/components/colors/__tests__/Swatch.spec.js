import React from 'react'
import { render, fireEvent } from 'react-testing-library'
import Swatch from '../Swatch'
import { ThemeProvider } from 'react-jss'
import theme from '../../../data/theme'

jest.useFakeTimers()

describe('Swatch', () => {
  it('renders correctly', () => {
    const mockProps = {
      color: { name: 'name', hex: '#fff' },
      classes: {},
    }
    const component = render(
      <ThemeProvider theme={theme}>
        <Swatch {...mockProps} />
      </ThemeProvider>
    )
    expect(component).toMatchSnapshot()
  })

  it('renders "Copied!" when correctly copied', () => {
    // mock document as not implemented by jsdom (https://jestjs.io/docs/en/manual-mocks#mocking-methods-which-are-not-implemented-in-jsdom)
    document.queryCommandSupported = jest.fn().mockImplementation(query => {
      return true
    })
    document.execCommand = jest.fn().mockImplementation(query => {
      return {
        matches: false,
        media: query,
        onchange: null,
        addListener: jest.fn(),
        removeListener: jest.fn(),
      }
    })
    const mockProps = {
      color: { name: 'name', hex: '#fff' },
      classes: {},
    }
    const { getByTestId, getByText } = render(
      <ThemeProvider theme={theme}>
        <Swatch {...mockProps} />
      </ThemeProvider>
    )
    const button = getByText(mockProps.color.hex)
    fireEvent.click(button)
    expect(getByTestId('swatch-text')).toHaveTextContent(/copied!/i)
    // TODO this doesn't actually test that the text changes...
    expect(setTimeout).toHaveBeenCalledTimes(1)
  })

  it('renders "Error!" if document.execCommand is not supported by browser', () => {
    // mock document as not implemented by jsdom (https://jestjs.io/docs/en/manual-mocks#mocking-methods-which-are-not-implemented-in-jsdom)
    document.queryCommandSupported = jest.fn().mockImplementation(query => {
      return false
    })
    document.execCommand = jest.fn().mockImplementation(query => {
      return {
        matches: false,
        media: query,
        onchange: null,
        addListener: jest.fn(),
        removeListener: jest.fn(),
      }
    })
    const mockProps = {
      color: { name: 'name', hex: '#fff' },
      classes: {},
    }
    const { getByTestId, getByText } = render(
      <ThemeProvider theme={theme}>
        <Swatch {...mockProps} />
      </ThemeProvider>
    )
    const button = getByText(mockProps.color.hex)
    fireEvent.click(button)
    expect(getByTestId('swatch-text')).toHaveTextContent(/error!/i)
    // TODO calls the above setTimeout also?
    expect(setTimeout).toHaveBeenCalledTimes(2)
  })
})
