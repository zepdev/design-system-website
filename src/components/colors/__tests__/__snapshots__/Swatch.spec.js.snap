// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Swatch renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="Swatch-root-0-1-1 Swatch-quarter-0-1-3"
      >
        <button
          class="ButtonBase-button-0-1-11 zep-button Swatch-button-0-1-4"
          style="background: rgb(255, 255, 255);"
        >
          <span
            class="Swatch-backdrop-0-1-5"
          />
          <span
            class="Swatch-backdropContent-0-1-6"
          >
            <svg
              aria-labelledby="title"
              class="Swatch-icon-0-1-7"
              height="32"
              viewBox="0 0 24 24"
              width="32"
              xmlns="http://www.w3.org/2000/svg"
            >
              <title
                id="title_copy"
              >
                Copy Icon
              </title>
              <path
                d="M0 0h24v24H0V0z"
                fill="none"
              />
              <path
                d="M16 1H4c-1.1 0-2 .9-2 2v14h2V3h12V1zm-1 4H8c-1.1 0-1.99.9-1.99 2L6 21c0 1.1.89 2 1.99 2H19c1.1 0 2-.9 2-2V11l-6-6zM8 21V7h6v5h5v9H8z"
                fill="currentColor"
              />
            </svg>
            <span
              class="Swatch-buttonText-0-1-8"
              data-testid="swatch-text"
            >
              #fff
            </span>
          </span>
        </button>
        <p
          class="Swatch-name-0-1-10 zep-typo--normal-2"
        >
          name
        </p>
        <p
          class="Swatch-text-0-1-9 zep-typo--normal-1"
        >
          HEX: #fff
        </p>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="Swatch-root-0-1-1 Swatch-quarter-0-1-3"
    >
      <button
        class="ButtonBase-button-0-1-11 zep-button Swatch-button-0-1-4"
        style="background: rgb(255, 255, 255);"
      >
        <span
          class="Swatch-backdrop-0-1-5"
        />
        <span
          class="Swatch-backdropContent-0-1-6"
        >
          <svg
            aria-labelledby="title"
            class="Swatch-icon-0-1-7"
            height="32"
            viewBox="0 0 24 24"
            width="32"
            xmlns="http://www.w3.org/2000/svg"
          >
            <title
              id="title_copy"
            >
              Copy Icon
            </title>
            <path
              d="M0 0h24v24H0V0z"
              fill="none"
            />
            <path
              d="M16 1H4c-1.1 0-2 .9-2 2v14h2V3h12V1zm-1 4H8c-1.1 0-1.99.9-1.99 2L6 21c0 1.1.89 2 1.99 2H19c1.1 0 2-.9 2-2V11l-6-6zM8 21V7h6v5h5v9H8z"
              fill="currentColor"
            />
          </svg>
          <span
            class="Swatch-buttonText-0-1-8"
            data-testid="swatch-text"
          >
            #fff
          </span>
        </span>
      </button>
      <p
        class="Swatch-name-0-1-10 zep-typo--normal-2"
      >
        name
      </p>
      <p
        class="Swatch-text-0-1-9 zep-typo--normal-1"
      >
        HEX: #fff
      </p>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllBySelectText": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getAllByValue": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getBySelectText": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "getByValue": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllBySelectText": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryAllByValue": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryBySelectText": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "queryByValue": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
